### ISSUE TITLE
Concatenate values from two JSON arrays by key

### ISSUE THREAD

**Issue by Keeline** [jqlang/jq#2373]
I am provided two JSON arrays from database fields.  Each has keys with numbers in quotes.  One is a set of questions (which varies across the data set) and the other is the corresponding answers.

`	{"1": "Property Type:", "2": "Credit Score", "3": "First Purchase", "4": "Purchase Situation", "5": "Property Used For", "6": "Purchase Price", "7": "Estimated Down Payment", "8": "Type Of Rate", "9": "Estimated Annual Household Income?", "10": "Employment Status", "11": "Bankruptcy Last 3 Years", "12": "Show Proof Of Income", "13": "Working Real Estate Agent", "14": "First Name", "15": "Last Name", "16": "Email", "17": "Phone"}
`	

`	{"1": "Single Family Home", "2": "Poor (539 or Lower)", "3": "Yes", "4": "Researching Options", "5": "Primary Residence", "6": "$80,001-$85,000", "7": "Zero Down", "8": "Fixed", "9": "$25,001-$50,000", "10": "Employed", "11": "no", "12": "yes", "13": "no", "14": "John", "15": "Smith", "16": "mail@exampple.com", "17": "2135551212"}
`

Ultimately, I would like to have the value from the first array paired up with its corresponding value.  I plan to provide this as a TSV with other data and each cell will contain the label and value separated by a colon like this (shown on separate lines for clarity):

Property Type: Single Family Home
Credit Score: Poor (539 or Lower)
"First Purchase: Yes
...

I have been tinkering with `jq` and trying to find the right way to resolve this without resorting to a bunch of nested loops.  The database has over a million rows.

Probably there is an easy way to do this and it is just not clear to me.

**Comment by pkoppstein** [comment]
The best place to ask usage questions such as this is stackoverflow.com with the jq tag.

If you want to try to find a robust solution to the problem, one strategy would be to start with:
```
jq -n '
  (input | map_values([.])) as $one
  | input as $two
  | reduce ($two|keys_unsorted[]) as $k2 ( $one;
      .[$k2] += [$two[$k2]] )
' 1.json 2.json
```

**Comment by Keeline** [comment]
Thank you.  It does seem like a good starting point.
James D. Keeline

    On Tuesday, November 16, 2021, 05:40:15 PM PST, pkoppstein ***@***.***> wrote:  
 
 


The best place to ask usage questions such as this is stackoverflow.com with the jq tag.

If you want to try to find a robust solution to the problem, one strategy would be to start with:
jq -n '
  (input | map_values([.])) as $one
  | input as $two
  | reduce ($two|keys_unsorted[]) as $k2 ( $one;
      .[$k2] += [$two[$k2]] )
' 1.json 2.json

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by itchyny** [comment]
Another option is to use JOIN filter.
```
❯ cat /tmp/1.json
{"1": "Property Type:", "2": "Credit Score", "3": "First Purchase", "4": "Purchase Situation", "5": "Property Used For", "6": "Purchase Price", "7": "Estimated Down Payment", "8": "Type Of Rate", "9": "Estimated Annual Household Income?", "10": "Employment Status", "11": "Bankruptcy Last 3 Years", "12": "Show Proof Of Income", "13": "Working Real Estate Agent", "14": "First Name", "15": "Last Name", "16": "Email", "17": "Phone"}

❯ cat /tmp/2.json
{"1": "Single Family Home", "2": "Poor (539 or Lower)", "3": "Yes", "4": "Researching Options", "5": "Primary Residence", "6": "$80,001-$85,000", "7": "Zero Down", "8": "Fixed", "9": "$25,001-$50,000", "10": "Employed", "11": "no", "12": "yes", "13": "no", "14": "John", "15": "Smith", "16": "mail@exampple.com", "17": "2135551212"}

❯ jq --slurpfile ks /tmp/1.json --slurpfile vs /tmp/2.json -n -r 'JOIN($vs[]; $ks[]|to_entries[]; .key; "\(.[0].value): \(.[1])")'
Property Type:: Single Family Home
Credit Score: Poor (539 or Lower)
First Purchase: Yes
Purchase Situation: Researching Options
Property Used For: Primary Residence
Purchase Price: $80,001-$85,000
Estimated Down Payment: Zero Down
Type Of Rate: Fixed
Estimated Annual Household Income?: $25,001-$50,000
Employment Status: Employed
Bankruptcy Last 3 Years: no
Show Proof Of Income: yes
Working Real Estate Agent: no
First Name: John
Last Name: Smith
Email: mail@exampple.com
Phone: 2135551212
```

**Comment by Keeline** [comment]
That is a useful example.  Thank you.  I will work to understand how these options work.  I have tested it with my content in files and see that it works.
As I mentioned, my JSON snippets are in MySQL fields so I will have them in a string.
I have worked out how to collapse the extra colon with a pipe to `tr -s ':'`.
I can also get the whole content on a single line separated by tabs with `tr '\n' \t'`.
However, I will still want to get the data in from strings.  I will explore that from the documentation but if there is an easy answer, I shall appreciate it.
Thank you again for helping me here when StackExchange is the preferred forum for this kind of question.
James D. Keeline


    On Tuesday, November 16, 2021, 11:02:39 PM PST, itchyny ***@***.***> wrote:  
 
 


Another option is to use JOIN filter.
❯ cat /tmp/1.json
{"1": "Property Type:", "2": "Credit Score", "3": "First Purchase", "4": "Purchase Situation", "5": "Property Used For", "6": "Purchase Price", "7": "Estimated Down Payment", "8": "Type Of Rate", "9": "Estimated Annual Household Income?", "10": "Employment Status", "11": "Bankruptcy Last 3 Years", "12": "Show Proof Of Income", "13": "Working Real Estate Agent", "14": "First Name", "15": "Last Name", "16": "Email", "17": "Phone"}

❯ cat /tmp/2.json
{"1": "Single Family Home", "2": "Poor (539 or Lower)", "3": "Yes", "4": "Researching Options", "5": "Primary Residence", "6": "$80,001-$85,000", "7": "Zero Down", "8": "Fixed", "9": "$25,001-$50,000", "10": "Employed", "11": "no", "12": "yes", "13": "no", "14": "John", "15": "Smith", "16": ***@***.***", "17": "2135551212"}

❯ jq --slurpfile ks /tmp/1.json --slurpfile vs /tmp/2.json -n -r 'JOIN($vs[]; $ks[]|to_entries[]; .key; "\(.[0].value): \(.[1])")'
Property Type:: Single Family Home
Credit Score: Poor (539 or Lower)
First Purchase: Yes
Purchase Situation: Researching Options
Property Used For: Primary Residence
Purchase Price: $80,001-$85,000
Estimated Down Payment: Zero Down
Type Of Rate: Fixed
Estimated Annual Household Income?: $25,001-$50,000
Employment Status: Employed
Bankruptcy Last 3 Years: no
Show Proof Of Income: yes
Working Real Estate Agent: no
First Name: John
Last Name: Smith
Email: ***@***.***
Phone: 2135551212

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
